### ISSUE TITLE
Build Failed: Cartographer on branch ros2 ros-foxy with colcon

### ISSUE THREAD

**Issue by ShivamPR21** [ros2/cartographer#18]
I was trying to get cartographer for ros2 on my system, but the build fails to give some errors, also I do not know about them much so here is my error log

```
--- stderr: cartographer                                         
/usr/bin/ld: CMakeFiles//usrcartographer.mapping.pose_graph_test.dir//bincartographer//ldmapping:/ pose_graph_test.cc.oCMakeFiles:/ cartographer.sensor.internal.voxel_filter_test.dirundefined/ cartographerreference/ sensorto/ internalsymbol/ voxel_filter_test.cc.o':_ZN7testing4TestC2Ev 'undefined
 /referenceusr /tobin /symbolld :' _ZN7testing4TestC2Ev/'home
//shivamusr//common_pkgsbin//googletest_2ld/:lib //libgtest.sohome:/ shivamerror/ common_pkgsadding/ googletest_2symbols/:lib /DSOlibgtest.so :missing  errorfrom  addingcommand  symbolsline:
 DSO missing from command line
collect2: error: ld returned 1 exit status
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/cartographer.sensor.internal.voxel_filter_test.dir/build.make:97: cartographer.sensor.internal.voxel_filter_test] Error 1
make[1]: *** [CMakeFiles/Makefile2:303: CMakeFiles/cartographer.sensor.internal.voxel_filter_test.dir/all] Error 2
make[2]: *** [CMakeFiles/cartographer.mapping.pose_graph_test.dir/build.make:97: cartographer.mapping.pose_graph_test] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:331: CMakeFiles/cartographer.mapping.pose_graph_test.dir/all] Error 2
/usr/bin/ld: CMakeFiles/cartographer.transform.transform_test.dir/cartographer/transform/transform_test.cc.o: undefined reference to symbol '_ZN7testing8internal24FormatMatcherDescriptionEbPKcRKSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS9_EE'
/usr/bin/ld: /home/shivam/common_pkgs/googletest_2/lib/libgmock.so: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/cartographer.transform.transform_test.dir/build.make:97: cartographer.transform.transform_test] Error 1
make[1]: *** [CMakeFiles/Makefile2:247: CMakeFiles/cartographer.transform.transform_test.dir/all] Error 2
/usr/bin/ld: CMakeFiles/cartographer.mapping.3d.range_data_inserter_3d_test.dir/cartographer/mapping/3d/range_data_inserter_3d_test.cc.o: undefined reference to symbol '_ZN7testing4TestC2Ev'
/usr/bin/ld: /home/shivam/common_pkgs/googletest_2/lib/libgtest.so: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/cartographer.mapping.3d.range_data_inserter_3d_test.dir/build.make:97: cartographer.mapping.3d.range_data_inserter_3d_test] Error 1
make[1]: *** [CMakeFiles/Makefile2:359: CMakeFiles/cartographer.mapping.3d.range_data_inserter_3d_test.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
---
Failed   <<< cartographer	[ Exited with code 2 ]
```

The build was succeeding for tag 2.1.0 but that is not compatible with  cartographer_ros
---
##### foxy
---

##### source
---
##### Ubuntu 20.04
---
#### build with the command:-
```colcon build --symlink-install --packages-up-to cartographer```

**Comment by clalancette** [comment]
Can you tell me which branch and commit you are on for the cartographer repository?  You have to be using the `ros2` branch here.

**Comment by ShivamPR21** [comment]
Yes I was using ros2 branch but before that I have tried to build on tag 2.0.0, 2.1.0 and release-latest unfortunately only 2.1.0 was able to build successfully but is incompatible with cartographer_ros[on branch ros2].

**Comment by clalancette** [comment]
Yes, you need to stick to the `ros2` branches for both this and cartographer_ros.

I will say that what you pasted above looks pretty weird.  There's a lot of jumbled output there, so it's kind of hard to see what is going on.

Unfortunately, I don't have much else to give you here.  This builds for me locally and on the buildfarm: http://build.ros2.org/view/Fbin_uF64/job/Fbin_uF64__cartographer__ubuntu_focal_amd64__binary/5/consoleFull .  This must be something different in your environment, but I don't have a good idea what it could be at the moment.

**Comment by goekce** [comment]
On Archlinux with GCC 10.2 I get similar errors:
```
/usr/bin/ld: CMakeFiles/cartographer.common.lockless_queue_test.dir/cartographer/common/lockless_queue_test.cc.o: undefined reference to symbol '_ZN7testing4Test5SetUpEv'
/usr/bin/ld: /usr/lib/libgtest.so.1.10.0: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
```
libgtest must be provided as a library during linking:
```
$ pwd
build/cartographer

$ cmake  -E cmake_link_script CMakeFiles/cartographer.common.lockless_queue_test.dir/link.txt --verbose=1
/usr/bin/c++ -O3 -DNDEBUG -rdynamic CMakeFiles/cartographer.common.lockless_queue_test.dir/cartographer/common/lockless_queue_test.cc.o -o cartographer.common.lockless_queue_test  libcartographer.a -lgmock_main -lpthread libcartographer_test_library.a libcartographer.a /usr/lib/libceres.so.1.14.0 -lglog -llua5.3 -lm /usr/lib64/libboost_iostreams.so.1.72.0 -lglog -lgflags -lcairo -lprotobuf -lpthread 
/usr/bin/ld: CMakeFiles/cartographer.common.lockless_queue_test.dir/cartographer/common/lockless_queue_test.cc.o: undefined reference to symbol '_ZN7testing4Test5SetUpEv'
/usr/bin/ld: /usr/lib/libgtest.so.1.10.0: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status

$ /usr/bin/c++ -O3 -DNDEBUG -rdynamic CMakeFiles/cartographer.common.lockless_queue_test.dir/cartographer/common/lockless_queue_test.cc.o -o cartographer.common.lockless_queue_test  libcartographer.a -lgmock_main -lpthread libcartographer_test_library.a libcartographer.a /usr/lib/libceres.so.1.14.0 -lglog -llua5.3 -lm /usr/lib64/libboost_iostreams.so.1.72.0 -lglog -lgflags -lcairo -lprotobuf -lpthread -lgtest
# links successfully
```
kunaltyagi has a CMake solution which works for me: https://github.com/cartographer-project/cartographer/issues/1611#issuecomment-560236725

**Comment by clalancette** [comment]
This should have been fixed upstream with https://github.com/cartographer-project/cartographer/commit/ae05658ff78c75e1be12c20408b1c996bb175695 (which is in this repository), and then further fixed in https://github.com/cartographer-project/cartographer/commit/1f69b83e8abc497c3c8207ae7748d1f06e9a82a0 (which is not).  So I *believe* that this would be fixed by moving to upstream.

And that is basically the goal here.  All of the changes in this repository have now been merged upstream, so this repository should be considered deprecated.  We can't quite switch to the upstream yet in ROS 2 Foxy, because the cartographer_ros port in https://github.com/ros2/cartographer_ros doesn't build against the newer cartographer.

All of that is to say is that you should make sure that https://github.com/cartographer-project/cartographer builds for you, and work with upstream to make sure that is the case.

**Comment by goekce** [comment]
I switched to `rolling` and tried compiling `cartographer_ros` with its dependencies. I replaced `ros2/cartographer` with `cartographer-project/cartographer` but as @clalancette suggested, `ros2/cartographer_ros` is not compatible with `cartographer-project/cartographer`. (I could compile `cartographer` but not `cartographer_ros`.)

I was about to replace `ros2/cartographer_ros` with `cartographer-project/cartographer_ros`, but saw the WIP https://github.com/cartographer-project/cartographer_ros/pull/1622 for `rolling`/`galactic`. I did not try it.
